File: Source/QText/Tray.vb
Imports System.Drawing

Friend Class Tray

    Private Shared mnxNotifyShow As New MenuItem("&Show")
    Private Shared mnxNotifyShowOnPrimary As New MenuItem("&Show on primary screen")
    Private Shared mnxNotifyExit As New MenuItem("E&xit")
    Private Shared notMain As New System.Windows.Forms.NotifyIcon
    Private Shared ReadOnly SyncRoot As New Object

	<System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Globalization", "CA1303:DoNotPassLiteralsAsLocalizedParameters", Justification:="This program is not intended to be localized.")> _
	Shared Sub New()
        App._form.CreateControl()
        App._form.Handle.GetType()

        mnxNotifyShow.DefaultItem = True
        AddHandler mnxNotifyShow.Click, AddressOf mnxNotifyShow_Click
        AddHandler mnxNotifyShowOnPrimary.Click, AddressOf mnxNotifyShowOnPrimary_Click
        AddHandler mnxNotifyExit.Click, AddressOf mnxNotifyExit_Click

        notMain.ContextMenu = New ContextMenu(New MenuItem() {mnxNotifyShow, mnxNotifyShowOnPrimary, New MenuItem("-"), mnxNotifyExit})
        notMain.Icon = Global.Medo.Resources.ManifestResources.GetIcon("QText.App.ico", 16, 16)
        notMain.Text = Global.Medo.Reflection.EntryAssembly.Title
        AddHandler notMain.MouseClick, AddressOf notMain_MouseClick
        AddHandler notMain.MouseDoubleClick, AddressOf notMain_MouseDoubleClick
    End Sub


    Public Shared Sub Init()
    End Sub

    Public Shared Sub Show()
        notMain.Visible = True
    End Sub

    Public Shared Sub Hide()
        notMain.Visible = False
    End Sub

    Private Shared Sub mnxNotifyShow_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        SyncLock Tray.SyncRoot
            App._form.Show()
            App._form.Activate()
        End SyncLock
    End Sub

    Private Shared Sub mnxNotifyShowOnPrimary_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        SyncLock Tray.SyncRoot
            App._form.Show()

            Dim priBounds As Rectangle = Screen.PrimaryScreen.WorkingArea
            Dim currBounds As Rectangle = App._form.Bounds
            Dim normalBounds As Rectangle
            If (App._form.WindowState = FormWindowState.Normal) Then
                normalBounds = App._form.Bounds
            Else
                normalBounds = App._form.RestoreBounds
            End If

            If (currBounds.Left >= priBounds.Left) AndAlso (currBounds.Right <= priBounds.Right) AndAlso (currBounds.Top >= priBounds.Top) AndAlso (currBounds.Bottom <= priBounds.Bottom) Then
            Else
                Dim oldState As FormWindowState = App._form.WindowState

                If oldState <> FormWindowState.Normal Then
                    App._form.WindowState = FormWindowState.Normal
                End If

                If (normalBounds.Width > priBounds.Width) Then
                    App._form.Width = priBounds.Width
                End If
                If (normalBounds.Left < priBounds.Left) Then
                    App._form.Left = priBounds.Left
                End If
                If (normalBounds.Right > priBounds.Right) Then
                    App._form.Left = priBounds.Right - normalBounds.Width
                End If

                If (normalBounds.Height > priBounds.Height) Then
                    App._form.Height = priBounds.Height
                End If
                If (normalBounds.Top < priBounds.Top) Then
                    App._form.Top = priBounds.Top
                End If
                If (normalBounds.Bottom > priBounds.Bottom) Then
                    App._form.Top = priBounds.Bottom - normalBounds.Height
                End If

                App._form.WindowState = oldState
            End If

            App._form.Activate()
        End SyncLock
    End Sub

	Private Shared Sub mnxNotifyExit_Click(ByVal sender As Object, ByVal e As System.EventArgs)
		Tray.Hide()
		Application.Exit()
	End Sub

    Private Shared Sub notMain_MouseClick(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
        If (e.Button = MouseButtons.Left) AndAlso (Settings.TrayOneClickActivation) Then
            Call mnxNotifyShow_Click(sender, New System.EventArgs)
        End If
    End Sub

	Private Shared Sub notMain_MouseDoubleClick(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
        If (e.Button = MouseButtons.Left) AndAlso (Not Settings.TrayOneClickActivation) Then
            Call mnxNotifyShow_Click(sender, New System.EventArgs)
        End If
	End Sub

End Class

File: Source/QText/Tray.cs
using System;
using System.Drawing;
using System.Windows.Forms;

namespace QText {
    internal class Tray : IDisposable {

        private NotifyIcon notMain = new NotifyIcon();
        public Form Form { get; private set; }


        private static readonly object SyncRoot = new object();
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Globalization", "CA1303:DoNotPassLiteralsAsLocalizedParameters", Justification = "This program is not intended to be localized.")]
        public Tray(Form form) {
            this.Form = form;

            this.Form.CreateControl();
            this.Form.Handle.GetType();

            notMain.Icon = Medo.Resources.ManifestResources.GetIcon("QText.Properties.App.ico", 16, 16);
            notMain.Text = Medo.Reflection.EntryAssembly.Title;

            notMain.MouseClick += delegate(object sender, MouseEventArgs e) {
                if ((e.Button == MouseButtons.Left) && (Settings.TrayOneClickActivation)) {
                    ShowForm();
                }
            };
            notMain.MouseDoubleClick += delegate(object sender, MouseEventArgs e) {
                if (e.Button == MouseButtons.Left) {
                    ShowForm();
                }
            };

            notMain.ContextMenu = new ContextMenu();
            notMain.ContextMenu.Popup += delegate(object sender, EventArgs e) {
                var showItem = new MenuItem("&Show") { DefaultItem = true };
                if (App.Hotkey.IsRegistered) { showItem.Text += "\t" + Helper.GetKeyString(App.Hotkey.Key); }
                showItem.Click += delegate(object sender2, EventArgs e2) {
                    ShowForm();
                };

                var showOnPrimaryItem = new MenuItem("Show on primary screen");
                showItem.Click += delegate(object sender2, EventArgs e2) {
                    ShowForm(true);
                };

                var exitItem = new MenuItem("E&xit");
                exitItem.Click += delegate(object sender2, EventArgs e2) {
                    this.Hide();
                    Application.Exit();
                };

                notMain.ContextMenu.MenuItems.Clear();
                notMain.ContextMenu.MenuItems.Add(showItem);
                notMain.ContextMenu.MenuItems.Add(showOnPrimaryItem);
                notMain.ContextMenu.MenuItems.Add(new MenuItem("-"));
                notMain.ContextMenu.MenuItems.Add(exitItem);
            };
        }


        public void Show() {
            notMain.Visible = true;
        }

        public void Hide() {
            notMain.Visible = false;
        }

        public void ShowForm(bool showOnPrimary = false) {
            lock (Tray.SyncRoot) {
                this.Form.Show();
                if (this.Form.WindowState == FormWindowState.Minimized) { this.Form.WindowState = FormWindowState.Normal; }

                if (showOnPrimary) {
                    Rectangle priBounds = Screen.PrimaryScreen.WorkingArea;
                    Rectangle currBounds = this.Form.Bounds;
                    Rectangle normalBounds = default(Rectangle);
                    if ((this.Form.WindowState == FormWindowState.Normal)) {
                        normalBounds = this.Form.Bounds;
                    } else {
                        normalBounds = this.Form.RestoreBounds;
                    }

                    if (!((currBounds.Left >= priBounds.Left) && (currBounds.Right <= priBounds.Right) && (currBounds.Top >= priBounds.Top) && (currBounds.Bottom <= priBounds.Bottom))) {
                        FormWindowState oldState = this.Form.WindowState;

                        if (oldState != FormWindowState.Normal) {
                            this.Form.WindowState = FormWindowState.Normal;
                        }

                        if ((normalBounds.Width > priBounds.Width)) {
                            this.Form.Width = priBounds.Width;
                        }
                        if ((normalBounds.Left < priBounds.Left)) {
                            this.Form.Left = priBounds.Left;
                        }
                        if ((normalBounds.Right > priBounds.Right)) {
                            this.Form.Left = priBounds.Right - normalBounds.Width;
                        }

                        if ((normalBounds.Height > priBounds.Height)) {
                            this.Form.Height = priBounds.Height;
                        }
                        if ((normalBounds.Top < priBounds.Top)) {
                            this.Form.Top = priBounds.Top;
                        }
                        if ((normalBounds.Bottom > priBounds.Bottom)) {
                            this.Form.Top = priBounds.Bottom - normalBounds.Height;
                        }

                        this.Form.WindowState = oldState;
                    }
                }

                this.Form.Activate();
            }
        }

        internal void ShowBalloonOnMinimize() {
            if (Settings.ShowBalloonOnNextMinimize) {
                Settings.ShowBalloonOnNextMinimize = false;
                var text = "Program continues to run in background.";
                if (App.Hotkey.IsRegistered) {
                    text += "\n\nPress " + Helper.GetKeyString(App.Hotkey.Key) + " to show window again.";
                }
                notMain.ShowBalloonTip(0, "QText", text, ToolTipIcon.Info);
            }
        }


        #region Disposing

        public void Dispose() {
            this.Dispose(true);
        }

        protected void Dispose(bool disposing) {
            if (disposing) {
                notMain.Dispose();
            }
        }

        #endregion

    }
}
